internal/extproc/translator/openai_openai.go

```diff
@@ -0,0 +1,102 @@
+package translator
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log/slog"
+
+	corev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	extprocv3http "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/ext_proc/v3"
+	extprocv3 "github.com/envoyproxy/go-control-plane/envoy/service/ext_proc/v3"
+
+	"github.com/envoyproxy/ai-gateway/internal/apischema/openai"
+)
+
+// newOpenAIToOpenAITranslator implements [TranslatorFactory] for OpenAI to OpenAI translation.
+func newOpenAIToOpenAITranslator(path string, l *slog.Logger) (Translator, error) {
+	if path == "/v1/chat/completions" {
+		return &openAIToOpenAITranslatorV1ChatCompletion{l: l}, nil
+	} else {
+		return nil, fmt.Errorf("unsupported path: %s", path)
+	}
+}
+
+// openAIToOpenAITranslatorV1ChatCompletion implements [Translator] for /v1/chat/completions.
+type openAIToOpenAITranslatorV1ChatCompletion struct {
+	defaultTranslator
+	l             *slog.Logger
+	stream        bool
+	buffered      []byte
+	bufferingDone bool
+}
+
+// RequestBody implements [RequestBody].
+func (o *openAIToOpenAITranslatorV1ChatCompletion) RequestBody(body *extprocv3.HttpBody) (
+	headerMutation *extprocv3.HeaderMutation, bodyMutation *extprocv3.BodyMutation, override *extprocv3http.ProcessingMode, modelName string, err error,
+) {
+	var req openai.ChatCompletionRequest
+	if err := json.Unmarshal(body.Body, &req); err != nil {
+		return nil, nil, nil, "", fmt.Errorf("failed to unmarshal body: %w", err)
+	}
+
+	if req.Stream {
+		o.stream = true
+		override = &extprocv3http.ProcessingMode{
+			ResponseHeaderMode: extprocv3http.ProcessingMode_SEND,
+			ResponseBodyMode:   extprocv3http.ProcessingMode_STREAMED,
+		}
+	}
+	headerMutation = &extprocv3.HeaderMutation{
+		SetHeaders: []*corev3.HeaderValueOption{{Header: &corev3.HeaderValue{Key: "host", Value: "api.openai.com"}}},
+	}
```

Comment:
could we allow user to rewrite the host instead of force to api.openai.com ?

Reply:
good catch - actually this needs to be handled in a different way

Reply:
https://github.com/envoyproxy/ai-gateway/pull/27/commits/33c5c4183f7d8476e7cb688623da2e434a02f4c3